src/playground-tab-bar.ts

```diff
@@ -0,0 +1,376 @@
+    // activeIndex property doesn't initially take. This hack pokes the bar
+    // after render to make sure the active tab is really selected.
+    if (!this._tabBar) {
+      return;
+    }
+    await this._tabBar.updateComplete;
+    this._tabBar.activeIndex = -1;
+    this._tabBar.activeIndex = this._activeFileIndex;
+  }
+
+  private _onProjectFilesChanged = () => {
+    this._setNewActiveFile();
+    if (this._editor) {
+      this._editor.filename = this._activeFileName;
+    }
+    this.requestUpdate();
+  };
+
+  private _onTabActivated(event: CustomEvent<{index: number}>) {
+    const index = event.detail.index;
+    const name = this._files[index].name;
+    if (name !== this._activeFileName) {
+      this._activeFileName = name;
+      this._activeFileIndex = index;
+      if (this._editor) {
+        this._editor.filename = name;
+      }
+    }
+  }
+
+  private _onOpenMenu(
+    event: CustomEvent<{index: number; anchor: HTMLElement}>
+  ) {
+    const controls = this._fileSystemControls;
+    if (!controls) {
+      return;
+    }
+    controls.state = 'menu';
+    controls.filename = this._files[event.detail.index].name;
+    controls.anchorElement = event.detail.anchor;
+  }
+
+  private _onClickAddFile(event: Event) {
+    const controls = this._fileSystemControls;
+    if (!controls) {
+      return;
+    }
+    controls.state = 'newfile';
+    controls.anchorElement = event.target as HTMLElement;
+  }
+
+  private _onNewFile(event: CustomEvent<{filename: string}>) {
+    this._activeFileName = event.detail.filename;
+    // TODO(aomarks) We should focus the editor here. However,
+    // CodeMirror.focus() isn't working for some reason.
+  }
+
+  /**
+   * Whenever a file is created, deleted, or renamed, figure out what the best
+   * new active tab should be.
+   */
+  private _setNewActiveFile() {
+    if (this._files.length === 0) {
+      this._activeFileIndex = 0;
+      this._activeFileName = '';
+      return;
+    }
+
+    // Stay on the same filename if it's still around, even though its index
+    // might have changed.
+    if (this._activeFileName) {
+      const index = this._files.findIndex(
+        (file) => file.name === this._activeFileName
+      );
+      if (index >= 0) {
+        this._activeFileIndex = index;
+        return;
+      }
+    }
+
+    // The active file was deleted, and it was the final tab. Move to the next
+    // tab down.
+    if (this._activeFileIndex > this._files.length - 1) {
+      this._activeFileIndex = this._files.length - 1;
+      this._activeFileName = this._files[this._activeFileIndex].name;
+      return;
+    }
+
+    // The active file was deleted, and another file has now taken the tab
+    // position of the old one. Select that file.
+    this._activeFileName = this._files[this._activeFileIndex].name;
+  }
+}
+
+/**
+ * Internal element for tabs within <playground-tab-bar>.
+ *
+ * This is a subclass of <mwc-tab>. We subclass because <mwc-tab> only supports
```

Comment:
Don't worry about this, but I thought about it since you're subclassing and adding components... another possible UI here is an ⓧ control on the right of the tab and making the filename editable inline with the tab.

Reply:
I like the dropdown because it gives us a place for more actions in the future. E.g. I was thinking of "Duplicate" and "Format".